openwillis/measures/video/facial_emotion.py

```diff
@@ -5,19 +5,159 @@
+
+    file = open(measure_path)
+    measures = json.load(file)
+    return measures
+
+def bb_dict_to_bb_list(bb_dict):
+    """
+    Convert a bounding box dictionary to a bounding box list.
+    Args:
+        bb_dict (dict): A dictionary containing bounding box coordinates with keys 'bb_x', 'bb_y', 'bb_w', and 'bb_h'.
+    Returns:
+        list: A nested list representing the bounding box in the format 
+              [[[bb_x, bb_y, bb_x + bb_w, bb_y + bb_h, 1]]], where 1 is the face confidence.
+    """
+
+    return [[[
+        bb_dict['bb_x'],
+        bb_dict['bb_y'],
+        bb_dict['bb_x'] + bb_dict['bb_w'],
+        bb_dict['bb_y'] + bb_dict['bb_y'],
+        1# this is to formatt bb_list to be compatible with pyfeat (this is face confidence)
+        ]]]
+
+def get_faces(detector,frame,bb_dict,threshold=.95):
+    '''
+    detect faces in frame if bb_dict is empty else use bb_dict
+
+    arguments:
+    detector: pyfeat detector object
+    frame: frame to detect faces in
+    bb_dict: dictionary with bounding box coordinates
+    threshold: threshold for face detection
+
+    returns:
+    faces: list of faces detected in frame
+    '''
+    if len(bb_dict.keys()):
+        faces = bb_dict_to_bb_list(bb_dict)
+    else:
+        faces = detector.detect_faces(
+                frame,
+                threshold=.95,
+            )
+    return faces
+
+def mouth_openness(
+    landmarks,
+    upper_lip_lmks=[61, 62, 63],
+    lower_lip_lmks=[65, 66, 67]
+):
+    """
+    ------------------------------------------------------------------------------------------------------
+    This function calculates the average distance between the upper and lower lip landmarks to measure the
+    openness of the mouth.
+
+    Parameters:
+    ..........
+    landmarks: numpy array
+        An array containing the facial landmarks detected by the facial landmark detector.
+    upper_lip_lmks: list
+        A list containing the indices of the landmarks that make up the upper lip.
+    lower_lip_lmks: list
+        A list containing the indices of the landmarks that make up the lower lip.
+    
+    Returns:
+    ..........
+    lmk_dist: float
+        The average distance between the upper and lower lip landmarks.
+    ------------------------------------------------------------------------------------------------------
+    """
+
+    upper_lip = landmarks[upper_lip_lmks]
+    lower_lip = landmarks[lower_lip_lmks]
+
+    lmk_dist = []
+    for (upper_lip_x, upper_lip_y), (lower_lip_x, lower_lip_y) in zip(upper_lip, lower_lip):
+        lmk_dist.append(np.sqrt((upper_lip_x - lower_lip_x)**2 + (upper_lip_y - lower_lip_y)**2))
+    return np.mean(lmk_dist)
+
+
+def detect_emotions(detector, frame, emo_cols, bb_dict={},threshold=.95):
+
+    faces = get_faces(
+        detector,
+        frame,
+        bb_dict,
+        threshold=threshold
+    )
+
+    if len(faces[0])<1:
```

Comment:
Add check for frames that didn't have a bounding box

```suggestion
    if len(faces[0])<1 or np.isnan(faces[0][0][:4]).sum() == 4:
```